File: src/java/client/WaypointClient.java
package cst420.thread.client;

import javax.swing.*;
import java.io.*;
import javax.swing.event.*;
import javax.swing.text.html.*;
import java.awt.event.*;
import java.awt.*;
import java.util.*;
import java.util.*;
import java.util.Arrays;
import java.text.DecimalFormat;

import java.net.URL;
import org.json.JSONObject;
import org.json.JSONArray;
import org.json.JSONException;


/**
 * Purpose: Class acts a client which takes in user inputs and data, sends requests to the server
 * for data, and displays a collection of data.
 * 
 * @author Brandon Sleater
 * @version November 20, 2014
 */
public class WaypointClient extends WaypointGUI implements ActionListener, ItemListener {

  private static final boolean debugOn = true;
  public String lookupVal;

  public static String serviceURL;
  public JsonRpcRequestViaHttp server;
  public static int id = 0;


  public WaypointClient(String url) {

    try {

      serviceURL = url;

      try {
        server = new JsonRpcRequestViaHttp(new URL(serviceURL));
        distBearInGC.setText("Connected at: " + serviceURL);
      } catch (Exception ex) {
        System.out.println("Malformed URL " + ex.getMessage());
      }

      //Load any currently saved waypoints from the server into the GUI
      loadAllWaypoints();
     
      removeWPButt.addActionListener(this);
      addWPButt.addActionListener(this);
      modWPButt.addActionListener(this);
      getAddrButt.addActionListener(this);
      getLatLonButt.addActionListener(this);
      distBearButtGC.addActionListener(this);
      frWps.addItemListener(this);
      toWps.addItemListener(this);

      this.addWindowListener(new WindowAdapter() {
        @Override
        public void windowClosing(WindowEvent e) {
           System.exit(0);
        }
      });

      setVisible(true);
    } catch (Exception ex) {
      JOptionPane.showMessageDialog(this, "Exception: " + ex.getMessage());
    }
  }


  private String packageCalcCall(String operation, String params) {

    JSONObject jsonObj = new JSONObject();
  
    jsonObj.put("jsonrpc", "2.0");
    jsonObj.put("method", operation);
    jsonObj.put("id", ++id);
    
    String almost   = jsonObj.toString();
    String toInsert = ",\"params\":"+params;

    String begin = almost.substring(0, almost.length() - 1);
    String end   = almost.substring(almost.length() - 1);
    String ret   = begin + toInsert + end;

    return ret;
  }


  /**
   * Function is used to refresh the dropdown list data. If running with a server
   * that other people manipulate, you would want to do this to get any data they added
   */
  public void loadAllWaypoints() {

    try {

      frWps.removeAllItems();
      toWps.removeAllItems();
  
      String jsonStr = this.packageCalcCall("getNamesFromLibrary", "null");
      String temp    = server.call(jsonStr);

      JSONObject points = new JSONObject(temp);
      
      Boolean isStr = false;

      try {
        String test = points.getString("result");
        isStr = true;
      } catch (JSONException e) { }

      //Ensure its an array
      if (!isStr) {

        JSONArray list = points.getJSONArray("result");

        for (int i = 0; i < list.length(); i++) {

          String value = list.getString(i);

          frWps.addItem(value);
          toWps.addItem(value);
        }

        loadWaypoint(String.valueOf(frWps.getSelectedItem()));
      }
    } catch(Exception ex) {
      System.out.println("RPC Exception Loading Points: " + ex.getMessage());
    }
  }


  /**
   * Function is used to load a single waypoint into the GUI
   */
  public void loadWaypoint(String name) {

    try {

      String params = "[\""+ name +"\"]";
  
      String jsonStr = this.packageCalcCall("getJSONPointFromLibrary", params);
      String temp    = server.call(jsonStr);

      JSONObject point = new JSONObject(temp);

      JSONObject res = point.getJSONObject("result");

      if (res.length() > 0) {

        point = point.getJSONObject("result").getJSONObject(name);

        namIn.setText(name);
        latIn.setText(String.format("%.4f", point.getDouble("lat")));
        lonIn.setText(String.format("%.4f", point.getDouble("lon")));
        eleIn.setText(String.format("%.4f", point.getDouble("ele")));
      }
    } catch(Exception ex) {
      System.out.println("RPC Exception Load Waypoint: " + ex.getMessage());
    }
  }


  public void addWaypoint() {

    String name = namIn.getText();
    String lat  = String.format("%.4f", Double.parseDouble(latIn.getText()));
    String lon  = String.format("%.4f", Double.parseDouble(lonIn.getText()));
    String ele  = String.format("%.4f", Double.parseDouble(eleIn.getText()));

    try {
  
      String params  = "[\"" + name + "\"," + lat + "," + lon + "," + ele + "]";
      String jsonStr = this.packageCalcCall("addWaypoint", params);

      server.call(jsonStr);
      loadAllWaypoints();

      frWps.setSelectedItem(name);
      toWps.setSelectedItem(name);
    } catch(Exception ex) {
      System.out.println("RPC Exception Add Waypoint: " + ex.getMessage());
    }
  }


  public void modifyWaypoint() {

    String name = namIn.getText();
    String lat  = String.format("%.4f", Double.parseDouble(latIn.getText()));
    String lon  = String.format("%.4f", Double.parseDouble(lonIn.getText()));
    String ele  = String.format("%.4f", Double.parseDouble(eleIn.getText()));

    try {
  
      String params  = "[\"" + name + "\"," + lat + "," + lon + "," + ele + "]";
      String jsonStr = this.packageCalcCall("modifyWaypoint", params);
      server.call(jsonStr);

      params  = "null";
      jsonStr = this.packageCalcCall("getActive", params);
      
      String temp       = server.call(jsonStr);
      JSONObject result = new JSONObject(temp);
      String active     = result.getString("result");

      loadAllWaypoints();

      frWps.setSelectedItem(active);
      toWps.setSelectedItem(active);
    } catch(Exception ex) {
      System.out.println("RPC Exception: " + ex.getMessage());
    }
  }


  public void removeWaypoint() {
  
    try {
  
      String params  = "null";
      String jsonStr = this.packageCalcCall("getActive", params);
      String temp    = server.call(jsonStr);

      JSONObject result = new JSONObject(temp);
      String active     = result.getString("result");

      params  = "[\""+ active +"\"]";
      jsonStr = this.packageCalcCall("removeWaypoint", params);
      
      String empty      = server.call(jsonStr);
      result            = new JSONObject(empty);
      Boolean lastEntry = result.getBoolean("result");

      if (lastEntry) {
        latIn.setText("");
        lonIn.setText("");
        eleIn.setText("");
        namIn.setText("");
      }

      loadAllWaypoints();
    } catch(Exception ex) {
      System.out.println("RPC Exception: " + ex.getMessage());
    }
  }


  public void calculateDistGC() {

    try {
  
      String wayFrom = String.valueOf(frWps.getSelectedItem());
      String wayTo   = String.valueOf(toWps.getSelectedItem());

      String params  = "[\"" + wayFrom + "\",\"" + wayTo + "\"]";
      String jsonStr = this.packageCalcCall("calcDistBear", params);
      
      String temp       = server.call(jsonStr);
      JSONObject result = new JSONObject(temp);
      JSONArray calc    = result.getJSONArray("result");

      //Clean up the output, 3 decimals should be fine
      DecimalFormat df = new DecimalFormat("#.####");

      distBearInGC.setText(df.format(calc.getDouble(0)) + " miles at " + df.format(calc.getDouble(1)));
    } catch(Exception ex) {
      System.out.println("RPC Exception: " + ex.getMessage());
    }
  }


  public void debug() {
  
    try {
  
      String jsonStr = this.packageCalcCall("debugLibrary", "null");
      server.call(jsonStr);
    } catch(Exception ex) {
      System.out.println("RPC Exception: " + ex.getMessage());
    }
  }


  public void itemStateChanged(ItemEvent event) {
     
    if (event.getStateChange() == ItemEvent.SELECTED) {
      try { loadWaypoint((String)event.getItem()); } catch (Exception ee) { ee.printStackTrace(); }
    }
  }


  /**
   * Function handles GUI actions that send requests up to the server
   */
  public void actionPerformed(ActionEvent e) {

    if (e.getActionCommand().equals("Add")) {
      addWaypoint();
    } else if (e.getActionCommand().equals("Remove")) {
      removeWaypoint();
    } else if (e.getActionCommand().equals("Modify")) {
      modifyWaypoint();
    } else if (e.getActionCommand().equals("DistanceGC")) {
      calculateDistGC();
    } else if (e.getActionCommand().equals("GetLatLon")) {
      debug();
    } else if (e.getActionCommand().equals("GetAddr")) {
      debug();
    }
  }


  public static void main(String args[]) {

    try {

      String url = "http://localhost:8080";

      if (args.length > 0){
        url = args[0];
      }
            
      WaypointClient cjc = new WaypointClient(url);         
    } catch (Exception e) {
      System.out.println("Error @ Main!");
    }
  }
}

File: src/cpp/WaypointClient.cpp
#include "WaypointGUI.cpp"
#include "waypointstubclient.h"

#include <FL/Fl.H>
#include <FL/Fl_Window.H>
#include <FL/Fl_Button.H>
#include <FL/Fl_Output.H>
#include <FL/Fl_Text_Display.H>
#include <FL/Fl_Text_Buffer.H>
#include <FL/Fl_Input_Choice.H>
#include <FL/Fl_Multiline_Input.H>
#include <stdio.h>
#include <stdlib.h>
#include <iostream>
#include <jsonrpc/rpc.h>

/**
 * Purpose: Class acts a client object. Receives events from the GUI
 *          and communicates with the serverObj to push back data to the GUI
 * 
 * @author Brandon Sleater
 * @version November 20, 2014
 */

using namespace std;
using namespace jsonrpc;

class WaypointClient : public WaypointGUI {

  /**
   * Makes a call to the serverObj object to get the current batch of 
   * saved waypoints. It then will rebuild the dropdown names for both 
   * from and to. If there are no waypoints, it will display placeholders
   *
   * @params (WaypointClient * anInstance, bool isEmpty)
   *    -anInstance: pointer to client object
   *    -isEmpty: flag to determine the label value (if waypoint or placeholder)
   * @return void
   */
  static void rebuildDropdowns(WaypointClient * anInstance, bool isEmpty) {
    
    waypointstubClient server(new HttpClient(hostID));
    Json::Value names;

    Fl_Input_Choice * fromDD = anInstance->frWps;
    Fl_Input_Choice * toDD   = anInstance->toWps;

    fromDD->clear();
    toDD->clear();

    fromDD->add("From Waypoint");
    toDD->add("To Waypoint");

    //Get all the waypoints currently on the serverObj member
    try {
      names = server.getNamesFromLibrary();
    } catch (JsonRpcException e) { cout << e.what() << endl; }

    for (int i = 0; i < names.size(); ++i) {

      string temp = names[i].asString();

      fromDD->add(temp.c_str());
      toDD->add(temp.c_str());
    }

    if (!isEmpty) {

      string nextPoint;

      //Get the last loaded point
      try {
        nextPoint = server.getActive();
      } catch (JsonRpcException e) { cout << e.what() << endl; }

      loadPoint(anInstance, nextPoint);

      fromDD->value(nextPoint.c_str());
      toDD->value(nextPoint.c_str());
    } else {

      //Clear out the text fields
      loadPoint(anInstance);

      //Set the original placeholders
      fromDD->value(fromDD->menubutton()->menu()[0].label());
      toDD->value(toDD->menubutton()->menu()[0].label());
    } 
  }


  /**
   * Remove a waypoint. It will make sure that the user is 
   * removing an actual waypoint, in which case, it will send 
   * a request to the serverObj object to remove it. It then determines
   * if there are any waypoints left for repopulating the dropdowns correctly
   *
   * @params (Fl_Widget * w, void * userdata)
   *    -w: pointer to GUI object
   *    -userdata: pointer to client object
   * @return void
   */
  static void clickedRemoveWP(Fl_Widget * w, void * userdata) {
    
    waypointstubClient server(new HttpClient(hostID));
    WaypointClient * anInstance = (WaypointClient *)userdata;

    Fl_Input_Choice * fromDD = anInstance->frWps;
    Fl_Input_Choice * toDD   = anInstance->toWps;

    string fromDDVal = fromDD->value();
    string toDDVal   = toDD->value();

    if (fromDDVal != "From Waypoint" && toDDVal != "To Waypoint") {

      string point;
      string response;
      bool isEmpty = false;
      
      //Instead of trying to determine which dropdown was selected, just
      //get the last point we loaded
      try {

        response = server.removeWaypoint(point);
        point    = server.getActive();

        isEmpty = (response == "false") ? false : true;
      } catch (JsonRpcException e) { cout << e.what() << endl; }

      rebuildDropdowns(anInstance, isEmpty);
    } else {

      //Empty the text fields
      loadPoint(anInstance);  
    }
  }

  
  /**
   * Add a waypoint. Will act like an actual user application 
   * and recontact the serverObj to rebuild the dropdowns
   * "incase other users have added data"
   *
   * @params (Fl_Widget * w, void * userdata)
   *    -w: pointer to GUI object
   *    -userdata: pointer to client object
   * @return void
   */
  static void clickedAddWP(Fl_Widget * w, void * userdata) {
    
    waypointstubClient server(new HttpClient(hostID));

    WaypointClient * anInstance = (WaypointClient *)userdata;

    Fl_Input_Choice * fromWPChoice = anInstance->frWps;
    Fl_Input_Choice * toWPChoice   = anInstance->toWps;

    Fl_Input * theLat  = anInstance->latIn;
    Fl_Input * theLon  = anInstance->lonIn;
    Fl_Input * theEle  = anInstance->eleIn;
    Fl_Input * theName = anInstance->nameIn;
    
    string name(theName->value());
    string lat(theLat->value());
    string lon(theLon->value());
    string ele(theEle->value());

    double latDub = atof(lat.c_str());
    double lonDub = atof(lon.c_str());
    double eleDub = atof(ele.c_str());

    try {
      server.addWaypoint(name, latDub, lonDub, eleDub);
    } catch (JsonRpcException e) { cout << e.what() << endl; }

    //Pass false because we know the serverObj member is not empty
    rebuildDropdowns(anInstance, false);
  }


  /**
   * Modify a waypoint. In case the user
   * changes the name, we rebuild the dropdown
   *
   * @params (Fl_Widget * w, void * userdata)
   *    -w: pointer to GUI object
   *    -userdata: pointer to client object
   * @return void
   */
  static void clickedModifyWP(Fl_Widget * w, void * userdata) {
    
    waypointstubClient server(new HttpClient(hostID));
    WaypointClient * anInstance = (WaypointClient *)userdata;

    Fl_Input_Choice * fromWPChoice = anInstance->frWps;
    Fl_Input_Choice * toWPChoice   = anInstance->toWps;

    Fl_Input * theLat  = anInstance->latIn;
    Fl_Input * theLon  = anInstance->lonIn;
    Fl_Input * theEle  = anInstance->eleIn;
    Fl_Input * theName = anInstance->nameIn;
    
    string name(theName->value());
    string lat(theLat->value());
    string lon(theLon->value());
    string ele(theEle->value());

    double latDub = atof(lat.c_str());
    double lonDub = atof(lon.c_str());
    double eleDub = atof(ele.c_str());

    try {
      server.modifyWaypoint(name, latDub, lonDub, eleDub);
    } catch (JsonRpcException e) { cout << e.what() << endl; }

    rebuildDropdowns(anInstance, false);
  }


  /**
   * Grab data from both active waypoints
   * and calculate their distance and bearing
   *
   * @params (Fl_Widget * w, void * userdata)
   *    -w: pointer to GUI object
   *    -userdata: pointer to client object
   * @return void
   */
  static void clickedDistBearWP(Fl_Widget * w, void * userdata) {

    waypointstubClient server(new HttpClient(hostID));
    Json::Value calcVals;
    char result[40] = {}, dist[10], bear[10];

    WaypointClient * anInstance = (WaypointClient *)userdata;

    Fl_Input * distBearWP = anInstance->distBearIn;

    Fl_Input_Choice * fromWPChoice = anInstance->frWps;
    Fl_Input_Choice * toWPChoice   = anInstance->toWps;

    string fromWP(fromWPChoice->value());
    string toWP(toWPChoice->value());

    try {
      calcVals = server.calcDistBear(fromWP, toWP);
    } catch (JsonRpcException e) { cout << e.what() << endl; }

    sprintf(dist, "%2.2f", calcVals[0u].asDouble());
    sprintf(bear, "%2.2f", calcVals[1u].asDouble());

    strcat(result, dist);
    strcat(result, " miles at ");
    strcat(result, bear);
    strcat(result, " degrees");

    distBearWP->value(result);
  }


  /**
   * Used as middle function to determine
   * which dropdown was selected. Sends frWps flag
   *
   * @params (Fl_Widget * w, void * userdata)
   *    -w: pointer to GUI object
   *    -userdata: pointer to client object
   * @return void
   */
  static void selectedFrom(Fl_Widget * w, void * userdata) {
    selectedPoint(w, userdata, 0);
  }


  /**
   * Used as middle function to determine
   * which dropdown was selected. Sends toWps flag
   *
   * @params (Fl_Widget * w, void * userdata)
   *    -w: pointer to GUI object
   *    -userdata: pointer to client object
   * @return void
   */
  static void selectedTo(Fl_Widget * w, void * userdata) {
    selectedPoint(w, userdata, 1);
  }


  /**
   * Used as middle function to load a waypoint
   *
   * @params (Fl_Widget * w, void * userdata)
   *    -w: pointer to GUI object
   *    -userdata: pointer to client object
   *    -whichOne: 0 => fromWps, 1 => toWps
   * @return void
   */
  static void selectedPoint(Fl_Widget * w, void * userdata, int whichOne) {

    waypointstubClient server(new HttpClient(hostID));
    WaypointClient * anInstance = (WaypointClient *)userdata;

    Fl_Input_Choice * click = whichOne ? anInstance->toWps : anInstance->frWps;

    string selected(click->value());

    //Make sure they clicked a waypoint
    //else get the last loaded point and load it
    //(should mimic the current loaded point from not changing if placeholder is selected)
    if (selected != "From Waypoint" && selected != "To Waypoint") {
      loadPoint(anInstance, selected);
    } else {
      try {
        click->value(server.getActive().c_str());
      } catch (JsonRpcException e) { cout << e.what() << endl; }
    }
  }


  /**
   * Used to get the the values of a 
   * waypoint name to output onto the GUI
   *
   * @params (Fl_Widget * w, void * userdata)
   *    -anInstance: pointer to client object
   *    -nameArg: waypoint to load. If nothing 
   *              passed, sets textfields empty
   * @return void
   */
  static void loadPoint(WaypointClient * anInstance, string nameArg = "") {

    waypointstubClient server(new HttpClient(hostID));

    Fl_Input * theLat  = anInstance->latIn;
    Fl_Input * theLon  = anInstance->lonIn;
    Fl_Input * theEle  = anInstance->eleIn;
    Fl_Input * theName = anInstance->nameIn;

    if (!nameArg.empty()) {

      Json::Value point;

      try {
        point  = server.getJSONPointFromLibrary(nameArg);
      } catch (JsonRpcException e) { cout << e.what() << endl; }

      //Format our waypoint values (4 decimal places)
      char lat[10], lon[10], ele[10];

      sprintf(lat, "%4.4f", point[nameArg]["lat"].asDouble());
      sprintf(lon, "%4.4f", point[nameArg]["lon"].asDouble());
      sprintf(ele, "%4.4f", point[nameArg]["ele"].asDouble());

      theLat->value(lat);
      theLon->value(lon);
      theEle->value(ele);
      theName->value(nameArg.c_str());
    } else {

      //Clear 'em
      theLat->value("");
      theLon->value("");
      theEle->value("");
      theName->value("");
    }
  }


  /**
   * Used to display parameters and values
   * at current points throughout functions
   *
   * @params (string message)
   *    -message: concatenated text and numbers 
   * @return void
   */
  static void debug(string message) {
    if (debugOn) cout << message << endl;
  }


  /**
   * Used as a debug to display the current
   * contents of the waypoint library
   *
   * @params (Fl_Widget * w, void * userdata)
   *    -w: pointer to GUI object
   *    -userdata: pointer to client object
   * @return void
   */
  static void clickedGetAddrWP(Fl_Widget * w, void * userdata) {
    
    waypointstubClient server(new HttpClient(hostID));
    
    try {
      server.debugLibrary();
    } catch (JsonRpcException e) { cout << e.what() << endl; }
  }


  /**
   * Terminate the application
   *
   * @params (Fl_Widget * w, void * userdata)
   *    -w: pointer to GUI object
   *    -userdata: pointer to client object
   * @return void
   */
  static void clickedX(Fl_Widget * w, void * userdata) {
    exit(1);
  }


  public:
    static const bool debugOn = false;
    static string hostID;

    WaypointClient(string host, const char * name = 0) : WaypointGUI(name) {

      //Save the hostID
      hostID = host;

      waypointstubClient server(new HttpClient(hostID));

      bool isEmpty = true;
      string active;

      try {
        cout << "Connected to: " << server.serviceInfo() << endl;
        active = server.getActive();
        isEmpty = (active == "null") ? true : false;
      } catch (JsonRpcException e) { cout << e.what() << endl; }

      rebuildDropdowns(this, isEmpty);

      distBearButt->callback(clickedDistBearWP, (void *)this);
      removeWPButt->callback(clickedRemoveWP, (void *)this);
      addWPButt->callback(clickedAddWP, (void *)this);
      modWPButt->callback(clickedModifyWP, (void *)this);
      frWps->callback(selectedFrom, (void *)this);
      toWps->callback(selectedTo, (void *)this);

      //Debug
      getAddrButt->callback(clickedGetAddrWP);

      //Exit
      callback(clickedX);
    }
};

string WaypointClient::hostID;

int main(int argc, char * argv[]) {

  string host = "http://127.0.0.1:8080";

  if (argc > 1) {
    host = string(argv[1]);
  }
  
  WaypointClient wc(host, "Waypoint Browser");

  return (Fl::run());
}
